galsim/phase_psf.py

```diff
@@ -0,0 +1,1399 @@
+        # out to twice the diameter of the actual aperture in the pupil plane (completely
+        # independent of wavelength, struts, obscurations, GSparams, and so on!).  This corresponds
+        # to oversampling=1.0.  In fact, if we were willing to always use sinc interpolation, there
+        # would never be any reason to go beyond this.  In practice, we usually use a faster, but
+        # less accurate, quintic interpolant, which means we can benefit from improved sampling
+        # (oversampling > 1.0) in some cases, especially when we're *not* modeling an atmosphere
+        # which would otherwise tend to damp contributions at large k.
+        good_pupil_size = 2 * diam * oversampling
+
+        # For the pupil plane sampling interval, details like the obscuration and GSParams *are*
+        # important as they affect the amount of aliasing encountered.  (An Airy profile has an
+        # infinite extent in real space, so it *always* aliases at some level, more so with an
+        # obscuration than without.  The GSParams settings indicate how much aliasing we're
+        # willing to tolerate, so it's required here.)  To pick a good sampling interval, we first
+        # check if a `screen_list` argument was supplied.  If so, we check it's .stepK() method,
+        # which aggregates a good sampling interval from all of the wrapped PhaseScreens.  If
+        # `screen_list` isn't supplied, then we fall back to creating an obscured Airy GSProfile and
+        # using it's .stepK().
+        if screen_list is not None:
+            screen_list = galsim.PhaseScreenList(screen_list)
+            if lam is None:
+                raise ValueError("Wavelength `lam` must be specified with `screen_list`.")
+            stepk = screen_list.stepK(lam=lam, diam=diam, obscuration=obscuration,
+                                      gsparams=self._gsparams)
+        else:
+            # For Airy, pupil_plane_scale is independent of wavelength.  We could build an Airy with
+            # lam_over_diam=1.0 and then alter the `good_pupil_scale = ...` line below
+            # appropriately, but it's easier to just arbitrarily set `lam=500` if it wasn't set.
+            if lam is None:
+                lam = 500.0
+            airy = galsim.Airy(diam=diam, lam=lam, obscuration=obscuration,
+                               gsparams=self._gsparams)
+            stepk = airy.stepK()
+        good_pupil_scale = (stepk * lam * 1.e-9 * (galsim.radians / galsim.arcsec)
+                            / (2 * np.pi * pad_factor))
+
+        # Now that we have good candidate sizes and scales, we
+        if pupil_plane_im is not None:  # Use image of pupil plane
+            self._load_pupil_plane(pupil_plane_im, pupil_angle, pupil_plane_scale,
+                                   good_pupil_scale, good_pupil_size)
+        else:  # Use geometric parameters.
+            if pupil_plane_scale is None:
+                pupil_plane_scale = good_pupil_scale
```

Comment:
Correct me if I'm wrong, but it seems to me that if the user has accidentally suggested bad values for ```pupil_plane_scale``` and ```pupil_plane_size``` (e.g., values that lead to undersampling), the code will happily use them instead of the ones it has estimated.  Should we be flagging this with a warning?